enum.c

```diff
@@ -589,6 +589,127 @@ enum_collect(VALUE obj)
+
+    return Qnil;
+}
+
+static VALUE
+scan_op_i(RB_BLOCK_CALL_FUNC_ARGLIST(i, memop))
+{
+    struct MEMO *memo = MEMO_CAST(memop);
+
+    ENUM_WANT_SVALUE();
+
+    if (memo->flags == Qfalse) {
+        if (memo->v2 == Qundef) {
+            MEMO_V2_SET(memo, i);
+        }
+
+        rb_ary_push(memo->v1, memo->v2);
+        memo->flags = Qtrue;
+    }
+
+    const ID mid = SYM2ID(memo->u3.value);
+    MEMO_V2_SET(memo, rb_funcallv_public(memo->v2, mid, 1, &i));
+
+    rb_ary_push(memo->v1, memo->v2);
+
+    return Qnil;
+}
+
+/*
+ *  call-seq:
+ *     enum.scan(initial, sym) -> array
+ *     enum.scan(sym)          -> array
+ *     enum.scan(initial) { |memo, obj| block }  -> array
+ *     enum.scan          { |memo, obj| block }  -> array
+ *
+ *  Computes a prefix scan of the elements of the collection.
+ *  Produces a collection of cumulative results starts from the <i>initial</i> value by
+ *  processing processing each element of the input series.
+ *
+ *  If you specify a block, then for each element in <i>enum</i>
+ *  the block is passed an accumulator value (<i>memo</i>) and the element.
+ *  If you specify a symbol instead, then each element in the collection
+ *  will be passed to the named method of <i>memo</i>.
+ *  In either case, the result becomes the new value for <i>memo</i>.
+ *
+ *  If you do not explicitly specify an <i>initial</i> value for <i>memo</i>,
+ *  then the first element of collection is used as the initial value
+ *  of <i>memo</i>.
+ *
+ *
+ *     # Sum some numbers
+ *     (5..10).scan(:+)                               #=> [5, 10, 16, 23, 31, 40, 50]
+ *     # Same using a block and scan
+ *     (5..10).scan { |total, n| total + n }          #=> [5, 10, 16, 23, 31, 40, 50]
+ *     # Multiply some numbers
+ *     (5..10).scan(1, :*)                            #=> [1, 5, 30, 210, 1680, 15120, 151200]
+ *     # Same using a block
+ *     (5..10).scan(1) { |product, n| product * n }   #=> [1, 5, 30, 210, 1680, 15120, 151200]
+ *
+ */
+static VALUE
+enum_scan(int argc, VALUE *argv, VALUE obj)
+{
+    struct MEMO *memo;
+    VALUE init, op;
+    rb_block_call_func *iter = scan_i;
+    ID id;
+
+    switch (rb_scan_args(argc, argv, "02", &init, &op)) {
+        case 0:
+            init = Qundef;
+            if (!rb_block_given_p()) {
+                rb_raise(rb_eArgError, "no block given");
+            }
+            break;
+        case 1:
+            if (rb_block_given_p()) {
+                break;
+            }
+            id = rb_check_id(&init);
+            op = id ? ID2SYM(id) : init;
+            init = Qundef;
+            iter = scan_op_i;
+            break;
+        case 2:
+            if (rb_block_given_p()) {
+                rb_warning("given block not used");
+            }
+            id = rb_check_id(&op);
+            if (id) op = ID2SYM(id);
+            iter = scan_op_i;
+            break;
+    }
+
+    memo = MEMO_NEW(rb_ary_new(), init, op);
+    memo->flags = Qfalse;
```

Comment:
Never change `flags` in `struct MEMO`, as it becomes a freed object.

Reply:
Is it OK to use `memo->u3.cnt` to the same approach?

Reply:
You are using `u3` for `op` already.
I'd suggest that initializing with `Qfalse` instead of `rb_ary_new()` here and replacing it later in `_i` functions.

Reply:
Thanks, for the suggestion. I found a quite obvious solution to check if an array is empty. What do you think should I rename `scan` to `reflect`?

Reply:
The name `scan` has been rejected already.